concurrency/ServiceController.java

```diff
@@ -0,0 +1,1922 @@
+                Logging.exception("searchMessages", e, "RemoteException");
+                return 0;
+            }
+        } else {
+            throw new IllegalStateException("No service found");
+        }
+    }
+
+    /**
+     * Increase the load count for a given mailbox, and trigger a refresh.  Applies only to
+     * IMAP and POP mailboxes, with the exception of the EAS search mailbox.
+     *
+     * @param mailboxId the mailbox
+     */
+    public static void loadMoreMessages(final Context context, final long mailboxId) {
+        // FIXME: This call was runAsyncParallel, to reduce rejectedExecutionException
+        // the call is made serial. Refer: PIM-17503
+        EmailAsyncTask.runAsyncSerial(new Runnable() {
+            @Override
+            public void run() {
+                Mailbox mailbox = Mailbox.restoreMailboxWithId(context, mailboxId);
+                if (mailbox == null) {
+                    return;
+                }
+                if (mailbox.mType == Mailbox.TYPE_SEARCH) {
+                    try {
+                        searchMore(context, mailbox.mAccountKey);
+                    } catch (MessagingException e) {
+                        // Nothing to be done
+                    }
+                    return;
+                }
+                AccountSettingsUtils.increaseSyncWindow(context, mailbox);
+                ServiceController.startSync(context, mailboxId, ServiceController.SYNC_UI_REFRESH);
+            }
+        });
+    }
+
+    private static final String UNSET_WARN_FLAG_SELECT = Message.MAILBOX_KEY + "=?";
+
+    /**
+     * Unset attachment warning flags for all messages in outbox
+     */
+    public static void unsetAttachmentSizeWarnings(final Context context, final long accountId) {
+        // FIXME: This call was runAsyncParallel, to reduce rejectedExecutionException
+        // the call is made serial. Refer: PIM-17503
+        EmailAsyncTask.runAsyncSerial(new Runnable() {
+            @Override
+            public void run() {
+                final long outboxId = Mailbox.findMailboxOfType(context, accountId,
+                        Mailbox.TYPE_OUTBOX);
+                if (outboxId == Mailbox.NO_MAILBOX) {
+                    return;
+                }
+                ContentResolver resolver = context.getContentResolver();
+                resolver.update(EmailContent.Message.UNSET_WARN_FLAG_URI, null,
+                        UNSET_WARN_FLAG_SELECT, new String[] { Long.toString(outboxId) });
+            }
+        });
+    }
+
+    public static boolean isPushInProgress(Context context, long mailboxId) {
+        ServiceController controller = sInstance;
+        if (controller == null) {
+            // By definition, it's not... :-)
+            return false;
+        }
+        ConcurrentHashMap<Long, Long> map = controller.mManagedMailboxMap;
+        if (map.containsKey(mailboxId)) {
+            mailboxId = map.get(mailboxId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **1826 and 1827**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **1479**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).